MekHQ/src/mekhq/campaign/autoresolve/acar/phase/FiringPhase.java

```diff
@@ -0,0 +1,222 @@
+import mekhq.campaign.autoresolve.acar.SimulationManager;
+import mekhq.campaign.autoresolve.acar.action.Action;
+import mekhq.campaign.autoresolve.acar.action.ManeuverToHitData;
+import mekhq.campaign.autoresolve.acar.action.StandardUnitAttack;
+import mekhq.campaign.autoresolve.component.EngagementControl;
+import mekhq.campaign.autoresolve.component.Formation;
+import mekhq.campaign.autoresolve.component.FormationTurn;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class FiringPhase extends PhaseHandler {
+
+    public FiringPhase(SimulationManager gameManager) {
+        super(gameManager, GamePhase.FIRING);
+    }
+
+    @Override
+    protected void executePhase() {
+        while (getSimulationManager().getGame().hasMoreTurns()) {
+
+            var optTurn = getSimulationManager().getGame().changeToNextTurn();
+            if (optTurn.isEmpty()) {
+                break;
+            }
+            var turn = optTurn.get();
+
+            if (turn instanceof FormationTurn formationTurn) {
+                var player = getSimulationManager().getGame().getPlayer(formationTurn.playerId());
+
+                getSimulationManager().getGame().getActiveFormations(player)
+                    .stream()
+                    .filter(f -> f.isEligibleForPhase(getSimulationManager().getGame().getPhase())) // only eligible formations
+                    .findAny()
+                    .map(this::attack)
+                    .stream()
+                    .flatMap(Collection::stream)
+                    .forEach(this::standardUnitAttacks); // add engage and control action
+            }
+        }
+    }
+
+    private void standardUnitAttacks(AttackRecord attackRecord) {
+        var actingFormation = attackRecord.actingFormation;
+        var target = attackRecord.target;
+
+        var attacks = new ArrayList<Action>();
+
+        var actingFormationToHit = ManeuverToHitData.compileToHit(actingFormation);
+        var targetFormationToHit = ManeuverToHitData.compileToHit(target);
+
+        ASRange range = ASRange.LONG;
+        StandardUnitAttack.ManeuverResult manueverModifier = StandardUnitAttack.ManeuverResult.DRAW;
+        if (!actingFormation.isRangeSet(target.getId())) {
+            var actingFormationMos = Compute.rollD6(2).getMarginOfSuccess(actingFormationToHit);
+            var targetFormationMos = Compute.rollD6(2).getMarginOfSuccess(targetFormationToHit);
+
+            if (actingFormationMos > targetFormationMos) {
+                if (EngagementControl.OVERRUN.equals(actingFormation.getEngagementControl())
+                    && !actingFormation.isEngagementControlFailed()) {
+                    range = ASRange.SHORT;
+                } else {
+                    range = WeightedDoubleMap.of(
+                        ASRange.LONG, actingFormation.getStdDamage().L.damage,
```

Comment:
Can we confirm whether the below CodeQLs are false-positive or not, it's hard to tell when reviewing through GitHub

Reply:
thats a false positive, It Is incapable of understanding the flow of this function.